[2678] Add transform to fill holes and to filter

Spenhouet opened the pull request:
Fixes #2678 .

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

Spenhouet commented:
@Nic-Ma @wyli I finalized my pull request but the merge checks throw a "import scipy.ndimage (No module named 'scipy')." and I don't get why? I used the same optional import of scipy.ndimage as it is done in `monai.apps.pathology.utils`.
Any idea how to fix that?

Nic-Ma commented:
Hi @Spenhouet ,

Thanks for your quick PR.
For the CI test errors, please:
1. Add the unit test file to https://github.com/Project-MONAI/MONAI/blob/dev/tests/min_tests.py#L138 to skip min tests.
2. Update the `type-hints` according to error message: monai/transforms/post/array.py:336:13: error: Returning Any from function declared to return "ndarray[Any, Any]"  [no-any-return]

Thanks.

Spenhouet commented:
@Nic-Ma I don't know how to address the type hint regarding numpy. This does not seem to have been done anywhere in the repository so far. I did use the NdarrayTensor type which is already in use. This one is also not further typeable. So I really don't know how to resolve this?

Nic-Ma commented:
Hi @Spenhouet ,

Maybe you can change that line to:
```py
return np.asarray(np.where(np.isin(img, self.applied_labels), img, 0))
```

Thanks.

Spenhouet commented:
@Nic-Ma This resolves it for mypy but then Pylance complains:

> Expression of type "ndarray[Unknown, Unknown]" cannot be assigned to return type "NdarrayTensor@__call__"
  Type "ndarray[Unknown, Unknown]" cannot be assigned to type "NdarrayTensor@__call__"

But since you are not linting with Pylance... this would resolve the issue I guess. Feels just like this actually a mypy issue and should be reported as bug on their side.
I will commit it like this for now.

Spenhouet commented:
@Nic-Ma @ericspod I feel like the discussion on #2678 is not finished yet. As I see it, the pull request should rather implement the `fill_holes7` logic since it is faster. But @ericspod mentioned that maybe someone might implement a faster solution (I guess directly in C?). Not sure if this should stay open until then or if we should try to merge an implementation of `fill_holes7` first?

wyli reviewed (commented):
I think we can merge this initial version first and improve the performance later if needed. What do you think @Nic-Ma @ericspod? Also, 'filter' is a very generic name (and name of a python build-in function), perhaps we should get a more specific name for the transform.

Spenhouet commented:
@wyli How about `LabelFilter` or `ValueFilter`?

wyli commented:
Labelfilter sounds good to me

Spenhouet commented:
I now changed the implementation to the "growing" logic. It is way faster than the connected component one and it becomes much faster if the filling is only done for a subset of labels.

I implemented it so that it directly works on batches and does not need to iterate over batches.

The pull request should be reviewable now.

Spenhouet commented:
@Nic-Ma Something failed here but I can't see the logs of the pull request checks.

ericspod commented:
We should merge this version first once we've figured out the issue with the tests. I can't see the logs either so @Spenhouet if you update the branch with changes from dev this will trigger the tests again and hopefully you'll get an error log this time. 

The solution later down the line is to define a convolution operation using max rather than sum as its function or otherwise define binary morphology operators in a Pytorch-compatible way on GPU.

wyli commented:
there was an issue with the github action today https://www.githubstatus.com/incidents/7p1nnvkgh96y let me rerun this pipeline

Spenhouet commented:
@wyli Thanks for rerunning the pipeline. The checks passed now.

Nic-Ma commented on docs/source/transforms.rst:
```diff
@@ -356,6 +356,18 @@ Post-processing
     :special-members: __call__
 
+`LabelFilter`
+"""""""""""""""""""""""""""""""
```
Could you please help change the length of  """""""" to be same as the words? Same as other titles.

Nic-Ma commented on monai/transforms/post/array.py:
```diff
@@ -289,6 +292,144 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+            of shape [batch_size, num_channel, spatial_dim1[, spatial_dim2, ...]].
+        """
+        if isinstance(img, np.ndarray):
+            channel_axis = 1
```
Most of MONAI transforms work with `channel-first` data, so it should be `channel_axis = 0`, and you may need to also update the unit tests.

Thanks.

Spenhouet replied:
This is channel first. Dimension 0 is the batch dimension.

Nic-Ma replied:
Hi @Spenhouet ,

The transform should work with single channel-first image, without batch dim.
You can refer to the `KeepLargestConnectedComponent`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L257.

Thanks.

Spenhouet replied:
You are right. I'm a bit confused since I remember having copy pasted the shape definition including the batch part from somewhere but I can't find it in the doc anywhere. Was this in a previous version or did I mix something up? What ever. I will fix it. Thanks for the hint.

Spenhouet replied:
Just sad since I invested at least 30 min. in making the code work with batches.

Nic-Ma replied:
Hi @Spenhouet ,

Thanks for your quick update.
Yes, we used batch-first data for postprocessing transforms before, but changed the behavior since MONAI v0.6.
You can take a look at this doc:
https://github.com/Project-MONAI/MONAI/wiki/v0.5-to-v0.6-migration-guide#decollate-batch-first-tensor-to-list-of-channel-first-tensors
Now we decollate the batch to a list, then execute postprocessing transforms.

Thanks.

Spenhouet replied:
I could leave the logic in to make it compatible with one-hot encoded data. Only the `applied_labels` filter wouldn't work then. It would always fill holes for all labels. It is valid to just throw a `NotImplementedError` in case someone runs this with one-hot and supplies the `applied_labels`? For non one-hot I would perform the squeeze then.

Spenhouet replied:
I might actually have a solution to also make the `applied_labels` work for one-hot encoded data.
But I'm confused again. I can swear I did see a transform before that handled both cases - single channel and one-hot encoded data. Where did that go?

Nic-Ma replied:
Hi @Spenhouet ,

I think you can refer to the `KeepLargestConnectedComponent`, it can handle One-Hot:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L273

Thanks.

Spenhouet replied:
I believe there is a different one-hot handling needed here. But you will see it when I implemented it :)

Nic-Ma commented on monai/transforms/post/array.py:
```diff
@@ -289,6 +292,144 @@ def __call__(self, img: torch.Tensor) -> torch.Tensor:
+        Initialize the connectivity and limit the labels for which holes are filled.
+
+        Args:
+            applied_labels (Optional[Union[Iterable[int], int]], optional): Labels for which to fill holes. Defaults to None,
```
As you already put type-hints in the code, no need to add the type in doc-strings.
Thanks.

Spenhouet replied:
I can remove that but it is actually part of the doc standard to do that.

Spenhouet commented:
Interestingly the performance for one-hot encoded images is way better when iterating over the encoding instead of letting the `binary_dilation` do it via the right window / structure.

Implementation with windows / structure:

```python
structure = np.zeros((3, *[3] * spatial_dims))
structure[1, ...] = ndimage.generate_binary_structure(spatial_dims, connectivity or spatial_dims)

img_arr_filtered = img_arr[tuple(applied_labels), ...]
tmp = np.zeros(img_arr_filtered.shape, dtype=bool)
ndimage.binary_dilation(
    tmp,
    structure=structure,
    iterations=-1,
    mask=np.logical_not(img_arr_filtered),
    origin=0,
    border_value=1,
    output=tmp,
)
img_arr[tuple(applied_labels), ...] = np.logical_not(tmp)
```

Implementation with iteration:

```python
structure = ndimage.generate_binary_structure(spatial_dims, connectivity or spatial_dims)

for label in applied_labels:
    tmp = np.zeros(img_arr.shape[1:], dtype=bool)
    ndimage.binary_dilation(
        tmp,
        structure=None,
        iterations=-1,
        mask=np.logical_not(img_arr[label]),
        origin=0,
        border_value=1,
        output=tmp,
    )
    img_arr[label] = np.logical_not(tmp)
```
| | structure  | iteration |
| ---- | ---- | ----- |
| 2 label | 73s | 1.4s |
| all labels | 134s | 31s |

I will therefore implement the iteration.

Spenhouet commented:
I pushed the change and now let's see if the checks pass.

Nic-Ma commented on monai/transforms/utils.py:
```diff
@@ -732,6 +734,65 @@ def get_largest_connected_component_mask(img: torch.Tensor, connectivity: Option
+            output=tmp,
+        )
+        if is_one_hot:
+            img_arr[label] = np.logical_not(tmp)
```
Hi @wyli @rijobro , I think for the One-Hot case, we should suppose `applied_labels` as channel indices instead of pure label value, because One-Hot only has values (0, 1), what do you think?

Thanks.

Spenhouet replied:
@Nic-Ma This is how it is implemented, or are you asking if this is acceptable?

wyli replied:
yes, looks good to me

Nic-Ma commented:
/black

Nic-Ma reviewed (approved):
Looks good to me now.
Thanks so much for the great feature and implementation!
